src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +91,334 @@ data BlockRollbackResult
+    randomR (lo,hi) = runRand $ uniform [lo..hi]
+
+genBlockEvent ::
+       (Monad m, RandomGen g)
+    => BlockEventGenParams
+    -> StateT GenBlockEventState (RandT g m) (Maybe (BlockEvent' BlockIndex))
+genBlockEvent begp = do
+    gbes <- get
+    failure <- lift $ byChance (begp ^. begpFailureChance)
+    when failure $ put GbesExpectedFailure
+    rollback <- lift $ byChance (begp ^. begpRollbackChance)
+    case gbes of
+        GbesExpectedFailure -> return Nothing
+        GbesStartingState -> Just <$> case (failure, rollback) of
+            (True, True) -> do
+                -- Fail with rollback. In the starting state it's easy,
+                -- because any rollback will fail when we don't have any
+                -- blocks yet.
+                blockIndices <- genBlockIndices 0
+                return . BlkEvRollback $ BlockEventRollback
+                    { _berInput    = toNewestFirst blockIndices
+                    , _berOutValid = BlockRollbackFailure
+                    }
+            (True, False) -> do
+                -- Fail without rollback (with apply). In the starting state,
+                -- the only way to do this is to generate a sequence of blocks
+                -- invalid by itself.
+                blockIndices <- genBlockIndices 1
+                let
+                    blkZeroth     = BlockIndex 0 :| []
+                    blockIndices' =
+                        -- Those block indices are broken by construction
+                        -- because we append the 0-th block to the end.
+                        -- Sometimes we can violate the maximum bound on
+                        -- block count, but this is fine. See NOTE
+                        -- on reordering corner cases.
+                        over _OldestFirst (Smg.<> blkZeroth) blockIndices
+                return . BlkEvApply $ BlockEventApply
+                    { _beaInput    = blockIndices'
+                    , _beaOutValid = BlockApplyFailure
+                    }
+            (False, _) -> do
+                -- Succeed with or without rollback. Unfortunately, it's
+                -- impossible to successfully rollback any blocks when
+                -- there are no blocks, so we will generate a block apply
+                -- event in both cases.
+                blockIndices <- genBlockIndices 0
+                put (GbesBlockchain blockIndices)
+                return . BlkEvApply $ BlockEventApply
+                    { _beaInput    = blockIndices
+                    , _beaOutValid = BlockApplySuccess
+                    }
+        GbesBlockchain blockchain -> Just <$> case (failure, rollback) of
+            (True, True) -> do
+                -- Fail with rollback.
+                rft <- getRandom
+                case rft of
+                    RftRollbackExcess -> do
+                        -- Attempt to rollback the entire blockchain and then
+                        -- some more.
```

Comment:
What does this failure mean exactly?
Suppose we have blockchain `[0, 1, 2]`. We may generate rollback `[2, 1, 0, -1]`. In the end we will generate 4 blocks. Okay, I see, rollback of the `-1`-st block will be unsuccessful. Can you add this example here to make it clearer for readers?